lib/semian/memcached.rb

```diff
@@ -0,0 +1,67 @@
+require 'semian/adapter'
+require 'memcached'
+
+module Semian
+  module Memcached
+    include Semian::Adapter
+
+    class SemianError < ::Memcached::Error
+      def initialize(semian_identifier, *args)
+        super(*args)
+        @semian_identifier = semian_identifier
+      end
+    end
+
+    ResourceBusyError = Class.new(SemianError)
+
+    def semian_identifier
+      name = semian_options && semian_options[:name]
+      ["memcached", name].compact.join("_").to_sym
+    end
+
+    def raw_semian_options
+      # disable semian when host ejection is enabled.
+      return false if !!@options[:auto_eject_hosts]
+      @options[:semian].merge(circuit_breaker: false)
+    end
+
+    # patch all the methods that actually interacts with memcached, not their higher level abstractions such
+    # as #get or #cas.
+    %i(
+      set
+      add
+      increment
+      decrement
+      replace
+      append
+      prepend
+      delete
+      exist
+      single_get
+      single_cas
+      multi_get
+      multi_cas
+    ).each do |meth|
+      define_method(meth) do |*args|
```

Comment:
I think we should raise `unless Memcached.method_defined?(meth)` so that things fail in an obvious way if one of these private methods gets removed or renamed in the memcached gem.  That seems quite likely given that the upstream master branch is doing a big refactor to remove the use of swig to generate the C extension.

Reply:
Yep we could. TBH, we are not looking at moving to memcached@master, but actually switching over to a pure ruby client such as Dalli. There's still a bit of work to do in order to acheive this, but it's something we want to do somewhere in the near future.

Reply:
Well, there is also a memcached 1-0-stable branch.  Also, it doesn't look like the tests actually cover all these methods, so I still think it makes sense to add the simple runtime check.